gobblin-core/src/main/java/gobblin/writer/http/AbstractHttpWriterBuilder.java

```diff
@@ -1,32 +1,132 @@
+                                                .put(POOLING_CONN_MANAGER_MAX_PER_CONN, 2)
+                                                .build();
+    FALLBACK = ConfigFactory.parseMap(configMap);
+  }
 
-  private HttpClientBuilder httpClientBuilder =
+  @Getter private State state;
+  @Getter private HttpClientBuilder httpClientBuilder =
       HttpClientBuilder.create().disableCookieManagement().useSystemProperties();
-  // TODO Add an option for connection pooling
-  private HttpClientConnectionManager httpConnManager = new BasicHttpClientConnectionManager();
+
+  @Getter private HttpClientConnectionManager httpConnManager;
+  @Getter private long reqTimeOut;
+  @Getter private Optional<Logger> logger = Optional.absent();
+  @Getter private Optional<URI> svcEndpoint = Optional.absent();
+
+  /**
+   * For backward compatibility on how Fork creates writer, invoke fromState when it's called writeTo method.
+   * @param destination
+   * @return
+   */
+  @Override
+  public B writeTo(Destination destination) {
+    super.writeTo(destination);
+    fromState(destination.getProperties());
+    return typedSelf();
+  }
 
   public B fromState(State state) {
+    this.state = state;
     Config config = ConfigBuilder.create().loadProps(state.getProperties(), CONF_PREFIX).build();
     fromConfig(config);
     return typedSelf();
   }
 
   public B fromConfig(Config config) {
+    config = config.withFallback(FALLBACK);
+    RequestConfig requestConfig = RequestConfig.copy(RequestConfig.DEFAULT)
+                                               .setSocketTimeout(config.getInt(REQUEST_TIME_OUT_MS_KEY))
+                                               .setConnectTimeout(config.getInt(CONNECTION_TIME_OUT_MS_KEY))
+                                               .setConnectionRequestTimeout(config.getInt(CONNECTION_TIME_OUT_MS_KEY))
+                                               .build();
+
+    httpClientBuilder.setDefaultRequestConfig(requestConfig);
+
+    if (config.hasPath(STATIC_SVC_ENDPOINT)) {
+      try {
+        svcEndpoint = Optional.of(new URI(config.getString(AbstractHttpWriterBuilder.STATIC_SVC_ENDPOINT)));
+      } catch (URISyntaxException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    String connMgrStr = config.getString(HTTP_CONN_MANAGER);
+    switch (ConnManager.valueOf(connMgrStr.toUpperCase())) {
+      case BASIC:
+        httpConnManager = new BasicHttpClientConnectionManager();
+        break;
+      case POOLING:
+        PoolingHttpClientConnectionManager poolingConnMgr = new PoolingHttpClientConnectionManager();
+        poolingConnMgr.setMaxTotal(config.getInt(POOLING_CONN_MANAGER_MAX_TOTAL_CONN));
+        poolingConnMgr.setDefaultMaxPerRoute(config.getInt(POOLING_CONN_MANAGER_MAX_PER_CONN));
+        httpConnManager = poolingConnMgr;
+        break;
+      default:
+        throw new IllegalArgumentException(connMgrStr + " is not supported");
+    }
+    LOG.info("Using " + httpConnManager);
```

Comment:
Something like this "Using org.apache.http.impl.conn.BasicHttpClientConnectionManager@225a7aab"

Will change it to class name.